### ISSUE TITLE
Error when sending get requests, might be related to redirects.

### ISSUE THREAD

**Issue by Zwork101** [aio-libs/aiohttp#2479]
I'm sending a simple get request, 
```python
async with self.session.get(self.CONF_URL + '/conf', params=params, headers=headers) as resp:
``` 
And I'm looking to receive html back, but it raises this error:
```
Traceback (most recent call last):
  File "C:\Users\Zwork101\PycharmProjects\python-trade\pytrade\confirmations.py", line 73, in get_confirmations
    async with self.session.get(self.CONF_URL + '/conf', params=params, headers=headers) as resp:
  File "C:\Users\Zwork101\AppData\Local\Programs\Python\Python36-32\lib\site-packages\aiohttp\client.py", line 692, in __aenter__
    self._resp = yield from self._coro
  File "C:\Users\Zwork101\AppData\Local\Programs\Python\Python36-32\lib\site-packages\aiohttp\client.py", line 264, in _request
    ssl_context=ssl_context, proxy_headers=proxy_headers)
  File "C:\Users\Zwork101\AppData\Local\Programs\Python\Python36-32\lib\site-packages\aiohttp\client_reqrep.py", line 105, in __init__
    url2 = url.with_query(params)
  File "C:\Users\Zwork101\AppData\Local\Programs\Python\Python36-32\lib\site-packages\yarl\__init__.py", line 768, in with_query
    "should be str or int, got {!r}".format(v))
TypeError: Invalid variable type: mapping value should be str or int, got b'Some bytes here'
```
This is the first half of the function, where the problem lies
```python
    async def get_confirmations(self):
        confs = ''
        params = self._create_confirmation_params('conf')
        headers = {'X-Requested-With': 'com.valvesoftware.android.steam.community'}
        async with self.session.get(self.CONF_URL + '/conf', params=params, headers=headers) as resp:
            if 'incorrect Steam Guard codes.' in resp.text():
                raise ValueError('Invalid identity_secret')
            confs = await resp.text()
```
I'm using python 3.6.1, and I used to be using an older version of aiohttp with the same code, and got a different error, one related to redirects. Then, I did `pip install -U aiohttp`, and now I'm getting this. Using the client version of aiohttp.

**Comment by asvetlov** [comment]
Check your `params` variable.
The exception explicitly blames that you have a `b'Some bytes here'` bytes in the mapping.
Please convert these `bytes` to `str`.

**Comment by Zwork101** [comment]
That stopped the mapping issue, but I'm getting the error I was having before 
```
ValueError: Can redirect only to http or https
``` 
Is this the server's fault? If so, the requests module can work with it.

**Comment by webknjaz** [comment]
@Zwork101 post new traceback and `CONF_URL` value.

**Comment by asvetlov** [comment]
Yep, looks like the server redirects to non-http URL

**Comment by webknjaz** [comment]
@asvetlov does aiohttp process relative `Location: ` header correctly? That's the only chance of bug left in the question.

**Comment by Zwork101** [comment]
I'll get the back when I'm home. I just want to put it out their, that this error does not occur when using the requests modual

**Comment by webknjaz** [comment]
`requests` library uses lots of dirty hacks under the hood :)

**Comment by Zwork101** [comment]
```
future: <Task finished coro=<ConfManager.get_confirmations() done, defined at C:\Users\Zwork101\PycharmProjects\python-trade\pytrade\confirmations.py:69> exception=ValueError('Can redirect only to http or https',)>
Traceback (most recent call last):
  File "C:\Users\Zwork101\PycharmProjects\python-trade\pytrade\confirmations.py", line 73, in get_confirmations
    async with self.session.get(self.CONF_URL + '/conf', params=params, headers=headers) as resp:
  File "C:\Users\Zwork101\AppData\Local\Programs\Python\Python36-32\lib\site-packages\aiohttp\client.py", line 692, in __aenter__
    self._resp = yield from self._coro
  File "C:\Users\Zwork101\AppData\Local\Programs\Python\Python36-32\lib\site-packages\aiohttp\client.py", line 330, in _request
    'Can redirect only to http or https')
ValueError: Can redirect only to http or https
```
This is the error, which seems quite odd. This is the CONF_URL: `"https://steamcommunity.com/mobileconf"`

**Comment by asvetlov** [comment]
The problem not in `CONF_URL` but in redirection URL returned by server:
```
$ curl -v -L https://steamcommunity.com/mobileconf/conf 
> GET /mobileconf/conf HTTP/1.1
> Host: steamcommunity.com
> User-Agent: curl/7.55.1
> Accept: */*
> 
< HTTP/1.1 302 Moved Temporarily
< Server: Apache
< X-Frame-Options: DENY
< Content-Security-Policy: script-src 'self' 'unsafe-inline' 'unsafe-eval' https://steamcommunity-a.akamaihd.net/ https://api.steampowered.com/ *.google-analytics.com https://www.google.com https://www.gstatic.com https://apis.google.com; object-src 'none'; connect-src 'self' http://steamcommunity.com https://steamcommunity.com https://api.steampowered.com/ http://localhost:27060 http://store.steampowered.com/ https://store.steampowered.com/; frame-src 'self' steam: http://store.steampowered.com/ https://store.steampowered.com/ http://www.youtube.com https://www.youtube.com https://www.google.com https://sketchfab.com https://player.vimeo.com;
< Expires: Mon, 26 Jul 1997 05:00:00 GMT
< Cache-Control: no-cache
< Location: steammobile://lostauth
< Content-Length: 0
< Content-Type: text/html; charset=UTF-8
< Date: Fri, 10 Nov 2017 18:17:20 GMT
< Connection: keep-alive
< Set-Cookie: sessionid=68e7f98cc575a19527e6d288; path=/
< Set-Cookie: steamCountry=UA%7Cebe2ed9d922f56a02588109e660ceaf4; path=/
< 
* Connection #0 to host steamcommunity.com left intact
* Issue another request to this URL: 'steammobile://lostauth'
* Rebuilt URL to: steammobile://lostauth/
* Protocol "steammobile" not supported or disabled in libcurl
* Closing connection -1
curl: (1) Protocol "steammobile" not supported or disabled in libcurl
```
Long story short: returned redirection URL is `steammobile://lostauth/`
*aiohttp* doesn't know how to work with `steammobile` schema and what is message format.
Is it HTTP at all?
CURL is confused too.

**Comment by Zwork101** [comment]
Thank you so much! How would I "pause" it, and stop it from redirecting, and go to another URL?

**Comment by asvetlov** [comment]
Use `session.get(url, allow_redirects=False)`

**Comment by Zwork101** [comment]
Thanks so much, sorry for wasting your time

**Comment by asvetlov** [comment]
No problem, you are welcome!

**Comment by webknjaz** [comment]
@asvetlov 

> Is it HTTP at all?

It's a custom schema used by native app. That's what makes Android to suggest user to open `twitter://` links in native Twitter app or `slack://` in native Slack app (works on other OSs too, handled by `xdg-open` @ linux for example).

Never seen that combined with HTTP, headers but it 100% works in my Google Chrome under Gentoo Linux (suggesting to open it via `xdg-open`).

Other question would be: should `aiohttp` provide some interface/sugar for supporting it? Like "stop on unsupported schema". Although it's not HTTP, but it's still a valid URL.

**Comment by asvetlov** [comment]
@webknjaz a new exception along with implemented #2335 makes sense.
Would you create an issue with clear description?

**Comment by webknjaz** [comment]
Sure #2507
